cache/models/samples.js

```diff
@@ -1148,78 +1251,152 @@ module.exports = {
-     * If there are filters, we need to load more records so we can return the
-     * right number of records in case some get filtered out later.
+    const filterKeys = Object.keys(opts.filter);
+    const hasFilter = filterKeys.length > 0;
+    const hasOrder = opts.order && opts.order.length > 0;
+    const hasNameFilterOnly = filterKeys.length === 1 && opts.filter.name;
+
+    /**
+     * Apply following logic when options have no filter and no sorting/sorting
+     * only by name.
+     * Get samplesKeys from sample set.
+     * Sort if sorting default or by name
+     * Apply limit/offset
+     * Get sample hashes, apply attributes filter, cleanup and
+     * return all samples.
      */
-    const sortArgs = [constants.indexKey.sample, 'alpha'];
-    const hasFilters = Object.keys(opts.filter).length > 0;
-    if (!hasFilters) {
-      sortArgs.push('LIMIT', opts.offset, opts.limit);
+    const shouldGetAllKeys = !hasFilter &&
+      (!hasOrder || opts.order === ['name'] || opts.order === ['-name']);
+
+    if (shouldGetAllKeys) {
+      return redisClient.smembersAsync(sampleStore.constants.indexKey.sample)
+        .then((sampleKeys) => {
+          sampleKeys.sort();
+
+          if (opts.order === ['-name']) {
+            sampleKeys.reverse();
+          }
+
+          const filteredSampleKeys = modelUtils.applyLimitAndOffset(
+            opts, sampleKeys);
+          return getSamplesAndAspects(filteredSampleKeys);
+        })
+        .then((samplesAndAspects) => {
+          const { samples, sampAspectMap } =
+            assembleSampleAspects(samplesAndAspects);
+          return applyAttributesFilter(samples, opts,
+            sampAspectMap, req.method);
+        });
     }
 
-    return Promise.resolve()
-    .then(() => {
-      // If there is a name param with no wildcards, get the sample directly.
-      const nameFilter = opts.filter.name;
-      if (nameFilter && !nameFilter.includes('*')) {
-        return getOneSample(nameFilter)
-        .then(([samp, asp]) =>
-          samp && asp ? [samp, asp] : []
-        );
-      }
+    /**
+     * Name filter and no wildcards:
+     * If {filter: { name: 's1|a1'}}
+     *  get one sample
+     *
+     * If { filter: { name: 's1|a1, s2|a2'}}
+     *  get the samples with comma separated names
+     *  If default name order, sort the array of sample names
+     *  Apply limit and offset
+     *  Get samples
+     *
+     * Then,
+     * If order other than name, then sort and apply limit/offset
+     * apply attributes filter, cleanup and return
+     *
+     */
+    if (hasNameFilterOnly && !opts.filter.name.includes('*')) {
+      let nameFilterArr = opts.filter.name.split(',').map((item) =>
+        item.trim());
+      return Promise.resolve()
+        .then(() => {
+          if (nameFilterArr.length === 1) { // one sample
+            return getOneSample(nameFilterArr[0])
+              .then(([samp, asp]) => samp && asp ? [samp, asp] : []);
```

Comment:
For multiple samples, we have `assembleSampleAspects` which does same thing after we 
Why necessary: We get all the samples and corresponding aspects in one redis batch command. So, this is one way to handle the result.

Reply:
assembleSampleAspects doesn't check that both the sample and the aspect exist. Why do we need to do that here?

Reply:
oh you are talking about the check - yeah, this code was already there, so it just felt the right thing to do. Yes, i noticed we do not check the existence in assembleSampleAspects. Maybe we should. How do you think it should be?
Should we create an error, or should we skip the bad sample like what we are doing here?

Reply:
Looking into this more, turns out I actually added this line as part of the quick fix optimization for the case where a single name was specified. Now that I remember, I did this because otherwise it throws an error when a nonexistent sample is specified. This wasn't necessary in the case of multiple samples because the sample names come directly from the set of existing keys. However, with this change, you're also using the multiple samples logic with names specified as comma-separated values, so I think that would cause an error if one of those doesn't exist.

Can you add a test for this case and make sure we handle that case?